research/conditional/moe_layers/token_choice.py

```diff
@@ -223,37 +258,19 @@ def forward(self, x: torch.Tensor):
 
         output = torch.zeros_like(x)
 
+        # here we would like to multiply by gating
         with measure_time(self, "assign_tokens_to_output"):
             for i in range(self.n_experts):
-                output[indices_of_tokens_for_expert[i]] = experts_output[
+                output[indices_of_tokens_for_expert[i]] += experts_output[
                     i, : len(indices_of_tokens_for_expert[i])
-                ]
-
-        # multiply output by softmax values
-        with measure_time(self, "multiply_output_by_softmax"):
-            if self.use_einsum:
-                output = einsum(
-                    "n_tokens dmodel, n_tokens -> n_tokens dmodel", output, expert_gate
+                ] * masked_expert_gate[indices_of_tokens_for_expert[i], i].unsqueeze(
+                    dim=1
```

Comment:
thats some involved code. Is that decomposable into multiple steps, perhaps?

Reply:
I made it a bit more readabloe